shared/Microsoft.Extensions.ObjectMethodExecutor.Sources/ObjectMethodExecutor.cs

```diff
@@ -0,0 +1,453 @@
+            //     (object)getCustomAwaitable(),
+            //     getAwaiterFunc,
+            //     isCompletedFunc,
+            //     getResultFunc,
+            //     onCompletedFunc,
+            //     unsafeOnCompletedFunc);
+            var returnValueExpression = Expression.New(
+                _objectMethodExecutorAwaitableConstructor,
+                Expression.Convert(methodCall, typeof(object)),
+                Expression.Constant(getAwaiterFunc),
+                Expression.Constant(isCompletedFunc),
+                Expression.Constant(getResultFunc),
+                Expression.Constant(onCompletedFunc),
+                Expression.Constant(unsafeOnCompletedFunc, typeof(Action<object, Action>)));
+
+            var lambda = Expression.Lambda<MethodExecutorAsync>(returnValueExpression, targetParameter, parametersParameter);
+            return lambda.Compile();
+        }
+
+        private static bool IsAwaitable(
+            Type type,
+            out MethodInfo getAwaiterMethod,
+            out Type awaiterType,
+            out Type returnType,
+            out PropertyInfo isCompletedProperty,
+            out MethodInfo getResultMethod,
+            out MethodInfo onCompletedMethod,
+            out MethodInfo unsafeOnCompletedMethod)
+        {
+            // Based on Roslyn code: http://source.roslyn.io/#Microsoft.CodeAnalysis.Workspaces/Shared/Extensions/ISymbolExtensions.cs,db4d48ba694b9347
+
+            // Awaitable must have method matching "object GetAwaiter()"
+            getAwaiterMethod = type.GetRuntimeMethods().FirstOrDefault(m => 
+                m.Name.Equals("GetAwaiter", StringComparison.OrdinalIgnoreCase)
+                && m.GetParameters().Length == 0
+                && m.ReturnType != null);
+            if (getAwaiterMethod == null)
+            {
+                awaiterType = null;
+                isCompletedProperty = null;
+                onCompletedMethod = null;
+                unsafeOnCompletedMethod = null;
+                getResultMethod = null;
+                returnType = null;
+                return false;
+            }
+
+            awaiterType = getAwaiterMethod.ReturnType;
+
+            // Awaiter must have property matching "bool IsCompleted { get; }"
+            isCompletedProperty = awaiterType.GetRuntimeProperties().FirstOrDefault(p =>
+                p.Name.Equals("IsCompleted", StringComparison.OrdinalIgnoreCase)
+                && p.PropertyType == typeof(bool)
+                && p.GetMethod != null);
+            if (isCompletedProperty == null)
+            {
+                onCompletedMethod = null;
+                unsafeOnCompletedMethod = null;
+                getResultMethod = null;
+                returnType = null;
+                return false;
+            }
+
+            // Awaiter must implement INotifyCompletion
+            var implementsINotifyCompletion = awaiterType.GetInterfaces().Any(t => t == typeof(INotifyCompletion));
```

Comment:
Instead of getting the interfaces twice, just reuse the list in the call too detect `ICriticalNotifyCompletion`